src/rulesets/Base/mapreduce.jl

```diff
@@ -48,3 +48,82 @@ function rrule(
     end
     return y, sum_abs2_pullback
 end
+
+#####
+##### `prod`
+#####
+
+function rrule(::typeof(prod), x::AbstractArray{T}; dims=:) where {T<:CommutativeMulNumber}
+    y = prod(x; dims=dims)
+    function prod_pullback(dy)
+        x_thunk = if dims == (:)
+            InplaceableThunk(
+                @thunk(∇prod(x, dy, y)),  # This is usually y ./ x .* dy
+                dx -> ∇prod!(dx, x, dy, y)
+            )
+        elseif any(iszero, x)  # Only cases where ./x would give NaN
+            InplaceableThunk(
+                @thunk(∇prod_dims(dims, x, dy, y)),
+                dx -> ∇prod_dims!(dx, dims, x, dy, y)
+            )
+        else
+            InplaceableThunk(
+                @thunk(y ./ conj.(x) .* conj.(dy)),
+                dx -> dx .+= y ./ conj.(x) .* conj.(dy)
+            )
+        end
+        (NO_FIELDS, x_thunk)
+    end
+    return y, prod_pullback
+end
+
+function ∇prod_dims(dims, x, dy, y=prod(x; dims=dims))
+    T = promote_type(eltype(x), eltype(dy))
+    dx = fill!(similar(x, T), zero(T))
+    ∇prod_dims!(dx, dims, x, dy, y)
+    return dx
+end
+
+function ∇prod_dims!(dx, dims, x, dy, y)
+    iters = ntuple(d -> d in dims ? tuple(:) : axes(x,d), ndims(x))
+    @inbounds for ind in Iterators.product(iters...)
+        jay = map(i -> i isa Colon ? 1 : i, ind)
+        @views ∇prod!(dx[ind...], x[ind...], dy[jay...], y[jay...])
+    end
+    return dx
+end
```

Comment:
this feels like it should be using the broadcasting machinery tto determine the axes.
But I can't quite see how.

Reply:
Wait don't you need the parentheses `Colon()`?

Reply:
it is using `isa`
```
julia> (:) isa Colon
true
```

Reply:
Oof it's checking the type, not inserting a colon - misread it. Carry on.